File: 2025/day02-1.hs
import Distribution.Simple.Utils (safeTail)

type Range = (Int, Int)

solution :: String -> Int
solution = sum . concatMap (filter isInvalidId . unwrap . parseRange) . parse

unwrap :: Range -> [Int]
unwrap (start, end) = [start .. end]

isInvalidId :: Int -> Bool
isInvalidId n = left == right
  where
    digits = show n
    half = length digits `div` 2
    left = take half digits
    right = drop half digits

parseRange :: String -> Range
parseRange line = (read . takeWhile (/= '-') $ line, read . tail . dropWhile (/= '-') $ line)

parse :: String -> [String]
parse "" = []
parse content = takeWhile (/= ',') content : (parse . safeTail . dropWhile (/= ',') $ content)

main :: IO ()
main = do readFile "day02.dat" >>= print . solution

File: 2025/day02_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: str) -> Any:
    return sum(
        item
        for rng in (parse(block) for block in content.split(","))
        for item in rng
        if is_invalid(item)
    )


def is_invalid(item: int) -> bool:
    half = len(str(item)) // 2
    left = str(item)[:half]
    right = str(item)[half:]
    return left == right


def parse(block: str) -> range:
    start, end = block.split("-")
    return range(int(start), int(end) + 1)


def main() -> None:
    with open("day02.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()
